### ISSUE TITLE
miscompiling i32 overflow

### ISSUE THREAD

**Issue by programmerjake** [Rust-GCC/gccrs#404]
https://rust.godbolt.org/z/coK3hz5aE

```rust
pub fn f(a: i32) -> i32 {
    if a + 1 == -0x80000000i32 {
        1
    } else {
        0
    }
}
```
is miscompiled to:
```asm
__f:
        xorl    %eax, %eax
        ret
```

**Comment by programmerjake** [comment]
the rust frontend needs to enable `-fwrapv`

**Comment by philberty** [comment]
Thanks for this this is interesting I am getting a different result locally:

GIMPLE
```c
i32 f (const i32 a)
{
  i32 D.204;

  _1 = a + 1;
  if (_1 == -2147483648) goto <D.201>; else goto <D.202>;
  <D.201>:
  {
    RUSTTMP.1 = 1;
  }
  goto <D.203>;
  <D.202>:
  {
    RUSTTMP.1 = 0;
  }
  <D.203>:
  D.204 = RUSTTMP.1;
  return D.204;
}
```

```
	.file	"test.rs"
	.text
	.globl	__f
	.type	__f, @function
__f:
.LFB0:
	.cfi_startproc
	pushq	%rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	movq	%rsp, %rbp
	.cfi_def_cfa_register 6
	movl	%edi, -4(%rbp)
	movl	-4(%rbp), %eax
	addl	$1, %eax
	cmpl	$-2147483648, %eax
	jne	.L2
	movl	$1, %eax
	jmp	.L4
.L2:
	movl	$0, %eax
.L4:
	popq	%rbp
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
.LFE0:
	.size	__f, .-__f
	.ident	"GCC: (GNU) 11.0.1 20210325 (experimental)"
	.section	.note.GNU-stack,"",@progbits
```

**Comment by programmerjake** [comment]
did you enable optimizations? I used -O3

**Comment by philberty** [comment]
Ah thanks for that i missed that.

**Comment by philberty** [comment]
Could be something to think about when doing #389

**Comment by bjorn3** [comment]
Using `-fwrapv` wouldn't work with cross-language LTO I think. I think it is an all or nothing flag for the whole program, so `-fwrapv` would either get enabled for other languages that keep it UB to overflow like C (and thus regress performance) or not enable it for Rust and thus cause UB in safe code.

**Comment by programmerjake** [comment]
yeah, too bad gcc has it as a program-wide global variable instead of having it be a per-instruction flag like llvm's `nsw` and `nuw` flags. https://llvm.org/docs/LangRef.html#add-instruction

**Comment by tschwinge** [comment]
For reference <https://doc.rust-lang.org/stable/book/ch03-02-data-types.html#integer-overflow>.

**Comment by tschwinge** [comment]
I haven't investigated all the gory details ;-) yet, but I notice that a few other GCC front ends indeed are setting `opts->x_flag_wrapv = 1`, which would match Rust's _release mode_.  For _debug mode_, I suppose we'd synthesize the appropriate checking code?  (Or, can we directly use GCC's `-ftrapv` for that?)

Needs a ton of testcases, obviously.  :-)

**Comment by tschwinge** [comment]
Regarding merging of these options (LTO, etc.), GCC states that they're _merged conservatively for conflicting translation units_.  I haven't tried to wrap my head around what exactly that means.  ;-)

**Comment by tschwinge** [comment]
It'll be some work, but I don't see a reason why GCC can't also expose this in the type `TREE`s directly: for example, instead of `gcc/tree.h:TYPE_OVERFLOW_WRAPS` checking `flag_wrapv`, we'd instead set that up based on `flag_wrapv` when the respective type `TREE` is laid out, and `TYPE_OVERFLOW_WRAPS` then just read the appropriate flag bit.

Indeed that seems to be a cleaner way of doing it, but I haven't researched the reason for the current implementation.

The current _late evaluation_ supposably allows for different `-fwrapv`/`-fno-wrapv` on a per function basis, using the same type `TREE` -- need to consider such cases, too?

**Comment by tschwinge** [comment]
Eh, or rather the other way round: have different type `TREE`s matching the different semantics (like LLVM IR's flags), and then use the correct one depending on current `flag_wrapv`, etc.

**Comment by rguenth** [comment]
flag_wrapv is recorded per function.  Note there's the desire to reflect flag_wrapv in the IL itself via builtins or internal functions which would support specifying it per operation.  Just nobody has done this work.  GCC -fwrapv currently only supports
add, subtract, multiplication, negation and signed division for trapping operations.  Lowering per operation would already
be possible by using the __builtin_{,u}{add,sub,mul}* builtins and conditionally calling the abortion mechanism required.
Note there's no builtin for negation or signed division where the check can be implemented with an equality to the most
negative number.  Note the builtins are later lowered to internal functions (see builtins.c:fold_builtin_arith_overflow) which
are easier to deal with and possibly also easier to directly generate from the Rust frontend.

I would strongly recommend against using -fwrapv.

**Comment by philberty** [comment]
This overflow behaviour is managed via "#[rustc_inherit_overflow_checks]" which is put on the add trait. https://github.com/rust-lang/rust/blob/cb3c4ee7187b045683cb9b86135dbbb766471091/library/core/src/ops/arith.rs#L99

**Comment by bjorn3** [comment]
Note that plain binops like `+` where the arguments are known to be integers are directly lowered to addition in MIR without involving a trait call. The implementation of binop traits for integers and floats is only important for typechecking and in generic contexts. `#[rustc_inherit_overflow_checks]` makes the MIR for a function contain overflow checks. When codegening a function in a crate with overflow checks disabled, any overflow checks in the MIR are ignored even if `#[rustc_inherit_overflow_checks]` is not used. `#[rustc_inherit_overflow_checks]` is solely about adding overflow checks to the MIR even though the standard library is compiled with overflow checks disabled.
